runtime/platform.c

```diff
@@ -0,0 +1,276 @@
+/*   Copyright 2016 University of Cambridge                               */
+/*                                                                        */
+/*   All rights reserved.  This file is distributed under the terms of    */
+/*   the GNU Lesser General Public License version 2.1, with the          */
+/*   special exception on linking described in the file LICENSE.          */
+/*                                                                        */
+/**************************************************************************/
+#define CAML_INTERNALS
+
+#include <string.h>
+#include <unistd.h>
+#include <errno.h>
+#include <sys/time.h>
+#include "caml/platform.h"
+#include "caml/fail.h"
+#ifdef HAS_SYS_MMAN_H
+#include <sys/mman.h>
+#endif
+#ifdef _WIN32
+#include <windows.h>
+#endif
+
+/* Mutexes */
+
+void caml_plat_mutex_init(caml_plat_mutex * m)
+{
+  int rc;
+  pthread_mutexattr_t attr;
+  rc = pthread_mutexattr_init(&attr);
+  if (rc != 0) goto error1;
+  rc = pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_ERRORCHECK);
+  if (rc != 0) goto error2;
+  rc = pthread_mutex_init(m, &attr);
+  // fall through
+error2:
+  pthread_mutexattr_destroy(&attr);
+error1:
+  check_err("mutex_init", rc);
+}
+
+void caml_plat_assert_locked(caml_plat_mutex* m)
+{
+#ifdef DEBUG
+  int r = pthread_mutex_trylock(m);
+  if (r == EBUSY) {
+    /* ok, it was locked */
+    return;
+  } else if (r == 0) {
+    caml_fatal_error("Required mutex not locked");
+  } else {
+    check_err("assert_locked", r);
+  }
+#endif
+}
+
+void caml_plat_assert_all_locks_unlocked()
+{
+#ifdef DEBUG
+  if (lockdepth) caml_fatal_error("Locks still locked at termination");
+#endif
+}
+
+void caml_plat_mutex_free(caml_plat_mutex* m)
+{
+  check_err("mutex_free", pthread_mutex_destroy(m));
+}
+
+static void caml_plat_cond_init_aux(caml_plat_cond *cond)
+{
+  pthread_condattr_t attr;
+  pthread_condattr_init(&attr);
+#if defined(_POSIX_TIMERS) && \
+    defined(_POSIX_MONOTONIC_CLOCK) && \
+    _POSIX_MONOTONIC_CLOCK != (-1)
+  pthread_condattr_setclock(&attr, CLOCK_MONOTONIC);
+#endif
+  pthread_cond_init(&cond->cond, &attr);
+}
+
+/* Condition variables */
+void caml_plat_cond_init(caml_plat_cond* cond, caml_plat_mutex* m)
+{
+  caml_plat_cond_init_aux(cond);
+  cond->mutex = m;
+}
+
+void caml_plat_wait(caml_plat_cond* cond)
+{
+  caml_plat_assert_locked(cond->mutex);
+  check_err("wait", pthread_cond_wait(&cond->cond, cond->mutex));
+}
+
+int caml_plat_timedwait(caml_plat_cond* cond, int64_t until)
+{
+  struct timespec t;
+  int err;
+  if (until < 0) {
+    /* until < 0 has definitely timed out, long ago.
+       letting the code below run risks feeding negative tv_nsec
+       to timedwait, since (-1 % 1000000000) = -1. */
+    return 1;
+  }
+  t.tv_sec  = until / 1000000000;
+  t.tv_nsec = until % 1000000000;
+  err = pthread_cond_timedwait(&cond->cond, cond->mutex, &t);
+  if (err == ETIMEDOUT) {
+    return 1;
+  } else {
+    return 0;
+  }
+}
```

Comment:
This function doesn't seem to be used anywhere at the moment.  Should we keep it?  Operations with relative timeouts always make me nervous.  Real-time software prefers absolute timeouts (deadlines), and non-real-time software should not care about time at all.

Reply:
Having Condition.timed_wait in some form (whether with absolute or relative timestamps) in the future would be useful.

Currently timeouts can be implemented by using pipes (as done here https://github.com/xapi-project/xen-api/blob/1e1165f330c30ee6cb8e8f325af9fdd07474ca24/ocaml/xapi-idl/lib/scheduler.ml), but condition variables might be more efficient and avoid the problem with `select` on pipes (`select` is limited to 1024 FDs, and there is no `poll` directly in the stdlib, so by using `select` this code limits the total number of FDs of the main app to 1024 unless its pipes are opened "early enough")

AFAICT `pthread_cond_timedwait` uses absolute time though.

Reply:
I stand corrected w.r.t absolute time.

Here we're talking about "platform" condition variables, i.e. those that are used in the C code for the OCaml runtime system.  It seems you're talking about the condition variables provided by the Condition module in OCaml.  The latter don't use "platform" condition variables; they have their own implementation in `runtime/caml/sync.h`.

So, whether to have timed-wait available to OCaml code from the Condition module is orthogonal to my question here, namely that `caml_plat_timedwait` is not used anywhere in the runtime system and I wonder if it's useful.

Reply:
`caml_plat_timedwait` is no longer used and should be removed. They were used in the old domain synchronization API: https://github.com/ocaml-multicore/ocaml-multicore/blob/4.06.1%2Bmulticore/stdlib/domain.mli#L94-L102, but not anymore.

Reply:
Removed in https://github.com/ocaml/ocaml/pull/10831/commits/6aa1379ba5a4891e4eb76ad6baafcb02ecd616a8.